### ISSUE TITLE
[BUG] Rollout vs collector in training loop

### ISSUE THREAD

**Issue by matteobettini** [pytorch/rl#1929]
Hello,

I was recently trying out substituting a `rollout` to my Sync collector in the sampling stage of multiagent training.

This is due to the fact that I need to compute gradients of the reward function (which the collector does not allow)

However, even with no gradients, replacing a rollout batch to the collector batch completely impedes training in my case.

Both the rollout and the collection are happening on a single vectorized environement with everything on the same device.

I have made an example script of my problem in #1928 where basically just by substituting `tensordict_data` from a rollout to the `tensordict_data` from the collector compeltely impedes training performance in a very simple task.

I would like more insights on why this is the case or if there is something wrong I am doing that i missed.

Thanks

P.S. all shapes of data from collector and rollout are the same

**Comment by vmoens** [comment]
Not sure I understand the issue, can you try to pinpoint a little more what's happening (maybe with a gist?)

Collectors do not do the same thing as rollout, mainly from a reset perspective. There are other things that might cause the issue... We do test that rollout and iterating over the collector spit out the same data. 
Perhaps you can try to do the same, iterating over the collector and using rollout and compare the results key-by-key?

**Comment by matteobettini** [comment]
> Not sure I understand the issue, can you try to pinpoint a little more what's happening (maybe with a gist?)

I'll try. What is happening is that everything works but the training rewards collapse in the rollout case, so I suspect that it has something to do with an interaction with the policy.

> Collectors do not do the same thing as rollout, mainly from a reset perspective

I tested it locally in an environemnt that just spits done at the end of max_steps and it occurs also in this case (no resets)

> We do test that rollout and iterating over the collector spit out the same data.
Perhaps you can try to do the same, iterating over the collector and using rollout and compare the results key-by-key?

I'll try this! i suspect that it has something to do with the policy interaction

**Comment by vmoens** [comment]
To mimic the dataloader you should pass a tensordict to use instead of reset to rollout, have you tried that?

**Comment by matteobettini** [comment]
You mean change

```python
          tensordict_data = env.rollout(
              max_steps=cfg.env.max_steps,
              policy=policy,
              break_when_any_done=False,
              auto_reset=True,
          )
```
to

```python
          td_reset = env.reset()
          tensordict_data = env.rollout(
              max_steps=cfg.env.max_steps,
              policy=policy,
              break_when_any_done=False,
              auto_reset=False,
              tensordict=td_reset,
          )
```
this does not change the outcome

**Comment by matteobettini** [comment]
Is there a simple training script in torchrl where we can try replacing the collector with a rollout that auto resets and see if we can reproduce this? I'll be willing to give that a go.

**Comment by vmoens** [comment]
Yeah seems reasonable. Would doing this solve your problem? 
That would also mean adding an optional reset when a tensordict is passed as input.

**Comment by matteobettini** [comment]
> Would doing this solve your problem?

No actually this is unfortunately unrelated. But I think it is a needed change anyway as I do not really see a purpose in resetting but not having access to the data.

My problem is giving me some serious debugging headaches, but I will get to the bottom of it. Basically just using a loop like below instead of a collector makes the rewards decrease over training.

> That would also mean adding an optional reset when a tensordict is passed as input.

Not sure I understand this, but after that change using the rollout in a loop would look like

```python
    reset_data = None
    any_done = False
    for i in range(epochs):
        with set_exploration_type(ExplorationType.RANDOM) and torch.no_grad():
            if reset_data is not None:
                if any_done:
                    td_reset = env.reset(reset_data)
            else:
                td_reset = env.reset()

            tensordict_data = env.rollout(
                max_steps=cfg.env.max_steps,
                policy=policy,
                break_when_any_done=False,
                auto_reset=False,
                tensordict=td_reset,
            )
            reset_data = step_mdp(
                tensordict_data[..., -1],
                keep_other=True,
                exclude_action=False,
                exclude_reward=True,
                reward_keys=env.reward_keys,
                action_keys=env.action_keys,
                done_keys=env.done_keys,
            )
            any_done = _terminated_or_truncated(
                reset_data,
                full_done_spec=env.output_spec["full_done_spec"],
                key="_reset",
            )
```

**Comment by matteobettini** [comment]
O M G

```python
from tensordict.nn.probabilistic import interaction_type
from torchrl.envs.utils import (
    _terminated_or_truncated,
    ExplorationType,
    set_exploration_type,
    step_mdp,
)
import torch

if __name__ == "__main__":
    with set_exploration_type(ExplorationType.RANDOM):
        print(interaction_type())

    with set_exploration_type(ExplorationType.RANDOM) and torch.no_grad():
        print(interaction_type())

```
```bash
InteractionType.RANDOM
None
```

**Comment by skandermoalla** [comment]
@matteobettini the last comment you sent looks worrying to me :o. That's not desired behaviour right?

**Comment by matteobettini** [comment]
The right synatax is with a comma
```
with torch.no_grad(), set_exploration_type(ExplorationType.RANDOM):
```

all my life I have been wrongly using `and` (it gives no error ahaha)

**Comment by skandermoalla** [comment]
🤯
